File: backend/internal/handler/quest_handler.go
package handler

import (
	"log/slog"
	"net/http"

	"github.com/gin-gonic/gin"
	"github.com/kenyamamoto/pokelingual/backend/internal/domain"
	"github.com/kenyamamoto/pokelingual/backend/internal/service"
)

// QuestHandler handles HTTP requests for the quest flow.
type QuestHandler struct {
	questService *service.QuestService
	repo         domain.UserPokemonRepository
}

// NewQuestHandler creates a new QuestHandler.
func NewQuestHandler(questService *service.QuestService, repo domain.UserPokemonRepository) *QuestHandler {
	return &QuestHandler{
		questService: questService,
		repo:         repo,
	}
}

// NewQuest handles GET /quest/new to start a new quest with a random Pokemon.
func (h *QuestHandler) NewQuest(c *gin.Context) {
	uid := c.GetString("uid")

	resp, err := h.questService.NewQuest(c.Request.Context(), uid)
	if err != nil {
		handleError(c, err)
		return
	}

	c.JSON(http.StatusOK, resp)
}

// ScoreTranslation handles POST /quest/score to score a user's translation.
func (h *QuestHandler) ScoreTranslation(c *gin.Context) {
	uid := c.GetString("uid")

	var req struct {
		Translation string `json:"translation" binding:"required"`
	}
	if err := c.ShouldBindJSON(&req); err != nil {
		c.JSON(http.StatusBadRequest, gin.H{"error": "translation is required"})
		return
	}

	resp, err := h.questService.ScoreTranslation(c.Request.Context(), uid, req.Translation)
	if err != nil {
		handleError(c, err)
		return
	}

	c.JSON(http.StatusOK, resp)
}

// GuessName handles POST /quest/guess-name to check a Pokemon name guess.
func (h *QuestHandler) GuessName(c *gin.Context) {
	uid := c.GetString("uid")

	var req struct {
		Guess string `json:"guess" binding:"required"`
	}
	if err := c.ShouldBindJSON(&req); err != nil {
		c.JSON(http.StatusBadRequest, gin.H{"error": "guess is required"})
		return
	}

	resp, err := h.questService.GuessName(c.Request.Context(), uid, req.Guess)
	if err != nil {
		handleError(c, err)
		return
	}

	c.JSON(http.StatusOK, resp)
}

// AttemptCapture handles POST /quest/capture to attempt capturing the Pokemon.
func (h *QuestHandler) AttemptCapture(c *gin.Context) {
	uid := c.GetString("uid")

	resp, err := h.questService.AttemptCapture(c.Request.Context(), uid)
	if err != nil {
		handleError(c, err)
		return
	}

	if err := h.repo.UpsertEncounter(c.Request.Context(), uid, resp.PokemonID, resp.Score, resp.Captured); err != nil {
		slog.Error("failed to persist encounter", "error", err, "uid", uid, "pokemon_id", resp.PokemonID)
	}

	c.JSON(http.StatusOK, resp)
}

File: backend/src/handler/quest-handler.ts
import type { Request, Response } from "express";
import type { UserPokemonRepository } from "../domain/ports.js";
import type { QuestService } from "../service/quest-service.js";
import type { ErrorResponse } from "../../../shared/api-types/error.js";
import { ExternalServiceError } from "../domain/errors.js";
import { handleError } from "./error.js";

/** クエスト関連エンドポイント (出題・採点・名前推測・捕獲) を束ねるハンドラ。 */
export class QuestHandler {
  /**
   * @param questService クエストのドメインサービス。
   * @param repo ユーザの図鑑進捗リポジトリ。
   */
  constructor(
    private questService: QuestService,
    private repo: UserPokemonRepository,
  ) {}

  /**
   * GET /quest/new — 新しい出題ポケモンを返す。
   * @param req Express リクエスト。
   * @param res Express レスポンス。
   */
  newQuest = async (req: Request, res: Response) => {
    const userId = res.locals.userId as string;
    try {
      const resp = await this.questService.newQuest(userId);
      res.json(resp);
    } catch (err) {
      handleError(res, err, req.path);
    }
  };

  /**
   * POST /quest/score — ユーザの翻訳文を LLM で採点しスコアを返す。
   * @param req Express リクエスト。
   * @param res Express レスポンス。
   */
  scoreTranslation = async (req: Request, res: Response) => {
    const userId = res.locals.userId as string;
    const { translation } = req.body;
    if (!translation) {
      res.status(400).json({ error: "translation is required" } satisfies ErrorResponse);
      return;
    }
    try {
      const resp = await this.questService.scoreTranslation(userId, translation);
      res.json(resp);
    } catch (err) {
      handleError(res, err, req.path);
    }
  };

  /**
   * POST /quest/guess-name — ポケモン名の推測を判定し残り試行回数を返す。
   * @param req Express リクエスト。
   * @param res Express レスポンス。
   */
  guessName = (req: Request, res: Response) => {
    const userId = res.locals.userId as string;
    const { guess } = req.body;
    if (!guess) {
      res.status(400).json({ error: "guess is required" } satisfies ErrorResponse);
      return;
    }
    try {
      const resp = this.questService.guessName(userId, guess);
      res.json(resp);
    } catch (err) {
      handleError(res, err, req.path);
    }
  };

  /**
   * POST /quest/skip-guess — 名前当てをスキップして、ボールを確定する。
   * @param req Express リクエスト。
   * @param res Express レスポンス。
   */
  skipGuess = (req: Request, res: Response) => {
    const userId = res.locals.userId as string;
    try {
      const resp = this.questService.skipGuess(userId);
      res.json(resp);
    } catch (err) {
      handleError(res, err, req.path);
    }
  };

  /**
   * POST /quest/capture — 捕獲を試行し結果を永続化する。
   * @param req Express リクエスト。
   * @param res Express レスポンス。
   */
  attemptCapture = async (req: Request, res: Response) => {
    const userId = res.locals.userId as string;
    try {
      const resp = this.questService.attemptCapture(userId);
      try {
        await this.repo.upsertEncounter(userId, resp.pokemon_id, resp.score, resp.captured);
      } catch (err) {
        throw new ExternalServiceError("Firestore", err as Error);
      }
      res.json(resp);
    } catch (err) {
      handleError(res, err, req.path);
    }
  };
}
